File: day20.jl
using ResumableFunctions

@enum Direction begin
    NORTH
    EAST
    SOUTH
    WEST
end 

const SideType = Tuple{Vararg{Bool, 10}}

struct TileConfig
    flip_x::Bool
    flip_y::Bool
    rotate_count::Int
    tile_id::Int
    north_side::SideType
    south_side::SideType
    west_side::SideType
    east_side::SideType
end

const EdgeDictType = Dict{Tuple{Direction, SideType}, Array{TileConfig}}

function makeTile(id::Int, tile::Array{String})::TileConfig
    north_side = Tuple([char == '#' for char=tile[1]])
    south_side = Tuple([char == '#' for char=tile[end]])
    west_side = Tuple([line[1] == '#' for line=tile])
    east_side = Tuple([line[end] == '#' for line=tile])
    TileConfig(false, false, 0, id, north_side, south_side, west_side, east_side)
end

function transformTile(config::TileConfig, flip_x, flip_y, rotate_count)::TileConfig
    id, north, south, west, east = config.tile_id, config.north_side, config.south_side, config.west_side, config.east_side

    if flip_y
        west, east = east, west
        north = reverse(north)
        south = reverse(south)
    end

    if flip_x
        north, south = south, north
        west = reverse(west)
        east = reverse(east)
    end

    for i = 1:rotate_count
        north, east, south, west = east, reverse(south), west, reverse(north)
    end

    TileConfig(flip_x, flip_y, rotate_count, id, north, south, west, east)
end

function addOrAppend!(dict::EdgeDictType, key, value)
    if haskey(edge_dict, key)
        push!(edge_dict[key], value)
    else
        edge_dict[key] = [value]
    end
    return nothing
end

file = open("input.txt", "r")
lines = readlines(file)
close(file)

tiles = NamedTuple{(:id, :tile), Tuple{Int, Array{String}}}[]

for i in range(1, length(lines) - 1, step=12)
    push!(tiles, (id=parse(Int, lines[i][5:end-1]), tile=lines[i+1:i+10]))
end


edge_dict = EdgeDictType()
all_config = TileConfig[]

for (id, tile) in tiles
    tile = makeTile(id, tile)
    for (flip_x, flip_y) in Iterators.product((true, false), (true, false))
        for r = 0:2
            tile = transformTile(tile, flip_x, flip_y, r)
            push!(all_config, tile)
            addOrAppend!(edge_dict, (NORTH, tile.north_side), tile)
            addOrAppend!(edge_dict, (SOUTH, tile.south_side), tile)
            addOrAppend!(edge_dict, (WEST, tile.west_side), tile)
            addOrAppend!(edge_dict, (EAST, tile.east_side), tile)
        end
    end
end

@show length(edge_dict)

@resumable function diagonal_iter(n::Int)::Tuple{Int, Int}
    for i in range(n*2 - 1, 0, step=-1)
        for x in range(n-1, 0, step=-1)
            y = i - x
            if y >= 0 && y < n
                @yield (x + 1, y + 1)
            end
        end
    end
end

function find_tiles(start_config::TileConfig, edge_dict::EdgeDictType, diagonals::Array{Tuple{Int, Int}})
    used_tile_ids = Set{Int}([start_config.tile_id])
    options_not_taken = Tuple{Int, Array{TileConfig}}[]
    tile_grid = Array{Union{TileConfig, Nothing}}(nothing, 12, 12)
    tile_grid[12, 12] = start_config
    diagonal_index = 2
    while diagonal_index <= length(diagonals)
        x, y = diagonals[diagonal_index]
        if y == 12 || (y == 1 && x != 12 && x != 1)
            available = [tile for tile = get(edge_dict, (EAST, tile_grid[x+1, y].west_side), []) if !(tile.tile_id in used_tile_ids)]
        elseif x == 12 || (x == 1 && y != 12 && y != 1)
            available = [tile for tile = get(edge_dict, (NORTH, tile_grid[x, y+1].south_side), []) if !(tile.tile_id in used_tile_ids)]
        else
            available_east = Set{TileConfig}(tile for tile = get(edge_dict, (EAST, tile_grid[x+1, y].west_side), []) if !(tile.tile_id in used_tile_ids))
            available = [tile for tile = get(edge_dict, (NORTH, tile_grid[x, y+1].south_side), []) if tile in available_east && !(tile.tile_id in used_tile_ids)]
        end
        # backtrack
        if length(available) == 0
            while true
                if diagonal_index <= 2
                    return nothing
                end
                oldx, oldy = diagonals[diagonal_index - 1]
                wrong_pick, other_opts = pop!(options_not_taken)
                delete!(used_tile_ids, wrong_pick)
                tile_grid[oldx, oldy] = nothing
                if length(other_opts) > 0
                    new_pick = pop!(other_opts)
                    push!(used_tile_ids, new_pick.tile_id)
                    tile_grid[oldx, oldy] = new_pick
                    push!(options_not_taken, (new_pick.tile_id, other_opts))
                else
                    diagonal_index -= 1
                end
            end
        else
            pick_tile = pop!(available)
            tile_grid[x, y] = pick_tile
            push!(options_not_taken, (pick_tile.tile_id, available))
            push!(used_tile_ids, pick_tile.tile_id)
            diagonal_index += 1
        end
    end
    tile_grid
end

diagonals = collect(diagonal_iter(12))

for (i, tile_config) in enumerate(all_config)
    tile_grid = find_tiles(tile_config, edge_dict, diagonals)
    if tile_grid !== nothing
        print(tile_grid[1,1].tile_id * tile_grid[12,1].tile_id * tile_grid[1,12].tile_id * tile_grid[12,12].tile_id)
        exit(0)
    end
end

File: 2020/day20.py
import itertools
import re
import math
import copy
from collections import defaultdict

# # Part A

class TileConfiguration:
    def __init__(self, flip_x, flip_y, rotate_count, tile_id, north_side, south_side, west_side, east_side):
        self.flip_x = flip_x
        self.flip_y = flip_y
        self.rotate_count = rotate_count # counter-clockwise
        self.tile_id = tile_id
        self.north_side = north_side
        self.south_side = south_side
        self.west_side = west_side
        self.east_side = east_side 

        if flip_y:
            self.west_side, self.east_side = self.east_side, self.west_side
            self.north_side = self.north_side[::-1]
            self.south_side = self.south_side[::-1]
        
        if flip_x:
            self.north_side, self.south_side = self.south_side, self.north_side
            self.east_side = self.east_side[::-1]
            self.west_side = self.west_side[::-1]
        
        for i in range(rotate_count):
            self.north_side, self.east_side, self.south_side, self.west_side = self.east_side, self.south_side, self.west_side, self.north_side

    def __repr__(self):
        return self.__str__()

    def __str__(self):
        return f"Tile: {self.tile_id}. Flip X: {self.flip_x}, Flip Y: {self.flip_y}, Rotations: {self.rotate_count}, North: {self.north_side} South: {self.south_side} East: {self.east_side} West: {self.west_side}"

with open('input.txt', 'r') as f:
  input = [line.strip() for line in f]

tiles = []

for i in range(0, len(input), 12):
    tile_num = int(input[i][5:-1])
    tiles.append((tile_num, input[i+1:i+11]))

edge_dict = defaultdict(list)

for tile_num, tile in tiles:
    north_side = tile[0]
    south_side = tile[9]
    west_side = "".join([x[0] for x in tile])
    east_side = "".join([x[9] for x in tile])

    for flip_x, flip_y in itertools.product([True, False], repeat=2):
        for i in range(3):
            config = TileConfiguration(flip_x, flip_y, i, tile_num, north_side, south_side, west_side, east_side)
            edge_dict[(north_side, "north")].append(config)
            edge_dict[(south_side, "south")].append(config)
            edge_dict[(west_side, "west")].append(config)
            edge_dict[(east_side, "east")].append(config)
    

def diagonal_iter():
    for i in range((12)*2 - 1, -1, -1):
        for y in range(12-1, -1, -1):
            x = i - y
            if x >= 0 and x < 12:
                yield (y, x)

diagonals = list(diagonal_iter())

def find_tiles(config):
    used_tile_ids = set(config.tile_id)
    options_not_taken = [[False for x in range(12)] for y in range(12)]
    tile_grid = [[False for x in range(12)] for y in range(12)]
    diagonal_index = 1
    while diagonal_index < len(diagonals):
        x, y = diagonals[diagonal_index]
        diagonal_index += 1
        if x == 0 and y == 0:
            continue
        if x == 12 or x == 0:
            # find available sides
            available = [tile for tile in edge_dict[(tile_grid[x][y-1].north_side, "south")] if tile.tile_id not in used_tile_ids]
            # fail
            if len(available) == 0:
                diagonal_index -= 1
                while True:
                    oldx, oldy = diagonals[diagonal_index - 1]
                    wrong_pick, other_opts = options_not_taken[oldx][oldy]
                    used_tile_ids.remove(wrong_pick)
                    if len(other_opts) > 0:
                        new_pick = other_opts.pop()
                        used_tile_ids.add(new_pick.tile_id)
                        tile_grid[oldx][oldy] = new_pick
                        options_not_taken[oldx, oldy] = (new_pick.tile_id, other_opts)
                        break
                    else:
                        diagonal_index -= 1

            else:
                pick_tile = available[0]
                tile_grid[x][y] = pick_tile
                options_not_taken[x][y] = (pick_tile.tile_id, available[1:])
                used_tile_ids.add(pick_tile.tile_id)
        elif y == 12 or y == 0:
            # find available sides
            available = [tile for tile in edge_dict[(tile_grid[x-1][y].west_side, "east")] if tile.tile_id not in used_tile_ids]
            # fail
            if len(available) == 0:
                # TODO
                pass
            else:
                pick_tile = available[0]
                tile_grid[x][y] = pick_tile
                options_not_taken[x][y] = (pick_tile.tile_id, available[1:])
                used_tile_ids.add(pick_tile.tile_id)
        else:
            # find available sides
            available_east = set([tile for tile in edge_dict[(tile_grid[x-1][y].west_side, "east")] if tile.tile_id not in used_tile_ids])
            available_north = [tile for tile in edge_dict[(tile_grid[x][y-1].north_side, "south")] if tile.tile_id not in used_tile_ids]
            available = [tile for tile in available_north if tile in available_east]
            # fail
            if len(available) == 0:
                # TODO
                pass
            else:
                pick_tile = available[0]
                tile_grid[x][y] = pick_tile
                options_not_taken[x][y] = (pick_tile.tile_id, available[1:])
                used_tile_ids.add(pick_tile.tile_id)

    print(config)

starting_tile = edge_dict.values().__iter__().__next__()[0]
# find_tiles(starting_tile, set())

